File: userland/network/tcpdump/packetdat.lua
-- EteleOS: userland/network/tcpdump/packetdat.lua, time wirte: 2026/08/08
-- This file uses the Apache-2.0 license
-- Summarize TCP packets in an FTP trace.  Ported from packetdat.awk.
if io.stdout and io.stdout.setvbuf then io.stdout:setvbuf("no") end
local argv = {...}; if #argv == 0 then argv = arg or {} end; local packetsize = tonumber(argv[1]) or 512
local filename = tonumber(argv[1]) and argv[2] or argv[1]
local input = filename and assert(io.open(filename, "r")) or io.stdin
local function fields(line) local r = {}; for x in line:gmatch("%S+") do r[#r+1] = x end; return r end
local function stamp(s) local h,m,sec = s:match("^(%d+):(%d+):([%d.]+)"); return h*3600 + m*60 + sec end
local firstSend, lastSend, firstAck, lastAck, seqNo, timesSent, timesAcked = {}, {}, {}, {}, {}, {}, {}
local maxId, totalPackets, totalAcks = 0, 0, 0
for line in input:lines() do
    local f = fields(line)
    if not (f[5] or ""):match("[SR]") then
        local time = stamp(f[1])
        if f[6] ~= "ack" then
            local seq = tonumber((f[6] or ""):match("^(.-):")) or 0
            local id = math.floor(1.5 + (seq - 1) / packetsize)
            maxId = math.max(maxId, id)
            if not firstSend[id] then firstSend[id], seqNo[id] = time, seq end
            lastSend[id] = time; timesSent[id] = (timesSent[id] or 0) + 1; totalPackets = totalPackets + 1
        else
            local id = math.floor(1 + ((tonumber(f[7]) or 0) - 2) / packetsize)
            timesAcked[id] = (timesAcked[id] or 0) + 1
            if not firstAck[id] then firstAck[id] = time end
            lastAck[id] = time; totalAcks = totalAcks + 1
        end
    end
end
if input ~= io.stdin then input:close() end
print(string.format("# %s chunks.  %s packets sent.  %s acks.", maxId, totalPackets, totalAcks))
for i = maxId - 1, 1, -1 do
    while i > 0 and not firstAck[i] do firstAck[i], lastAck[i], i = firstAck[i + 1], firstAck[i + 1], i - 1 end
end
local tzero = firstSend[1] or 0
for i = 1, maxId do
    io.write(string.format("%d\t%d\t%.2f\t%.2f\t%.2f\t%.2f\t%d\t%d\n", i, seqNo[i] or 0,
        (firstSend[i] or 0) - tzero, (lastSend[i] or 0) - tzero,
        (firstAck[i] or 0) - tzero, (lastAck[i] or 0) - tzero,
        timesSent[i] or 0, timesAcked[i] or 0))
end

File: userland/network/tcpdump/packetdat.pl
# PeteleOS: userland/network/tcpdump/packetdat.pl, time wirte: 2026/08/26
# This file uses the Apache-2.0 license
#
# Converted from packetdat.awk to Perl.
#
# The original packetdat.awk carried no separate copyright/license
# header of its own; it is one of the small ftp-trace analysis scripts
# distributed alongside OpenBSD's tcpdump (see README in this
# directory). Its original comments are preserved below verbatim.
#
# we need to know (usual) packet size to convert byte numbers
# to packet numbers
#
# print out per-packet data in the form:
#  <packet #>
#  <start sequence #>
#  <1st send time>
#  <last send time>
#  <1st ack time>
#  <last ack time>
#  <# sends>
#  <# acks>
#
# usage: perl packetdat.pl [packetsize=NNN] tracedata

use strict;
use warnings;

my $packetsize = 0;

# emulate awk's "var=value" command-line assignment convention
my @files;
foreach my $arg (@ARGV) {
	if ($arg =~ /^packetsize=(\S+)$/) {
		$packetsize = $1 + 0;
	} else {
		push @files, $arg;
	}
}
@ARGV = @files;

$packetsize = 512 if $packetsize <= 0;

my %firstSend;
my %lastSend;
my %seqNo;
my %timesSent;
my %firstAck;
my %lastAck;
my %timesAcked;
my $maxId = 0;
my $totalPackets = 0;
my $totalAcks = 0;

while (my $line = <>) {
	chomp $line;
	my @f = split ' ', $line;
	next if defined $f[4] && $f[4] =~ /[SR]/;

	my @t = split /:/, $f[0];
	my $tim = $t[0] * 3600 + $t[1] * 60 + $t[2];

	if (!defined $f[5] || $f[5] ne 'ack') {
		my $i = index($f[5], ':');
		my $strtSeq = substr($f[5], 0, $i);
		my $id = 1.5 + ($strtSeq - 1) / $packetsize;
		$id = int($id);
		$maxId = $id if $maxId < $id;
		if (!$firstSend{$id}) {
			$firstSend{$id} = $tim;
			$seqNo{$id} = $strtSeq;
		}
		$lastSend{$id} = $tim;
		$timesSent{$id}++;
		$totalPackets++;
	} else {
		my $id = 1 + ($f[6] - 2) / $packetsize;
		$id = int($id);
		$timesAcked{$id}++;
		$firstAck{$id} = $tim if !$firstAck{$id};
		$lastAck{$id} = $tim;
		$totalAcks++;
	}
}

print "# $maxId chunks.  $totalPackets packets sent.  $totalAcks acks.\n";

# for packets that were implicitly acked, make the ack time
# be the ack time of next explicitly acked packet.
for (my $i = $maxId - 1; $i > 0; $i--) {
	while ($i > 0 && !$firstAck{$i}) {
		$lastAck{$i} = $firstAck{$i} = $firstAck{$i + 1};
		$i--;
	}
}

my $tzero = $firstSend{1} // 0;
for (my $i = 1; $i <= $maxId; $i++) {
	printf "%d\t%d\t%.2f\t%.2f\t%.2f\t%.2f\t%d\t%d\n",
		$i, $seqNo{$i} // 0,
		($firstSend{$i} // 0) - $tzero, ($lastSend{$i} // 0) - $tzero,
		($firstAck{$i} // 0) - $tzero, ($lastAck{$i} // 0) - $tzero,
		$timesSent{$i} // 0, $timesAcked{$i} // 0;
}
